diff --git a/kernel/include/ksu.h b/kernel/include/ksu.h
--- a/kernel/include/ksu.h
+++ b/kernel/include/ksu.h
@@ -8,12 +8,8 @@
 #define KERNEL_SU_VERSION KSU_VERSION
 #define KERNEL_SU_VERSION_TAG KSU_VERSION_TAG
 
 extern struct cred *ksu_cred;
-extern bool ksu_late_loaded;
 extern bool allow_shell;
-#ifdef MODULE
-extern bool ksu_bundled;
-#endif
 extern struct selinux_policy *backup_sepolicy;
 extern bool ksu_no_custom_rc;
 
